Cohort discussion settings and roles management API

lgp171188 opened the pull request:
This PR implements an alternative cohort discussion settings and cohort discussion roles management API for server-to-server communication. This is related to PR https://github.com/edx/edx-platform/pull/18975 PR which implements a few other API endpoints for managing cohorts. The endpoints support OAuth2 authentication and session authentication and build upon existing API endpoints for the purpose.

**Sandbox URL**: TBD - sandbox is being provisioned.

**Merge deadline**: None

**Testing instructions**:
* Log in to the LMS as a staff user.
* Create a test course and enable cohorts for it.
* Create some test users.
* Create a few cohorts and enroll the test users at random in the test course into the created cohorts.
* Enable the 'Course Discussions' feature for the course.
* Test and verify the course discussion settings endpoints using the format of the following example requests. More details are available in the view docstring for these endpoints.

        # Get the discussion settings for the given course.
        GET /api/discussion/v1/courses/{course_id}/settings

        # Example response
        {
            "always_divide_inline_discussions": false,
            "divided_inline_discussions": [],
            "divided_course_wide_discussions": [],
            "id": 1,
            "division_scheme": "cohort",
            "available_division_schemes": ["cohort"]
        }

        # Update the discussion settings for the given course.
        PATCH /api/discussion/v1/courses/{course_id}/settings
        {
            "always_divide_inline_discussions": true,
        }
        # Example response
        {
            "always_divide_inline_discussions": true,
            "divided_inline_discussions": [],
            "divided_course_wide_discussions": [],
            "id": 1,
            "division_scheme": "cohort",
            "available_division_schemes": ["cohort"]
        }

        # Get the member users for a given course and discussion role (URL encoded if necessary) - moderator, group moderator or community TA.
        GET /api/discussion/v1/courses/{course_id}/roles/{discussion rolename}

        # Example response
        {
            "results": [
                {
                    "username": "user1",
                    "email": "user1@example.com",
                    "first_name" "first name",
                    "last_name": "last name",
                    "group_name": "cohort1"
                },
                {
                    "username": "user2",
                    "email": "user2@example.com",
                    "first_name" "first name",
                    "last_name": "last name",
                    "group_name": "cohort2"
                }
            ]
        }

       # Update the discussion role membership of a user in the given course.
        POST /api/discussion/v1/courses/{course_id}/roles/{discussion rolename}

        {
            "user_id": "user1",
            "action": "revoke"
        }

        # Example response
        {
            "results": [
                {
                    "username": "user2",
                    "email": "user2@example.com",
                    "first_name" "first name",
                    "last_name": "last name",
                    "group_name": "cohort2"
                }
            ]
        }


Reviewers
- [ ] @viadanna 
- [ ] edX reviewer[s] TBD

edx-webhook commented:
Thanks for the pull request, @lgp171188! I've created [OSPR-2647](https://openedx.atlassian.net/browse/OSPR-2647) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

lgp171188 commented:
@natabene, since the PR was a WIP at the time of its creation, the description did not have enough details. I have added more details now.

@viadanna, this is now ready for your review.

viadanna commented on lms/djangoapps/discussion_api/views.py:
```diff
@@ -568,3 +586,248 @@ def post(self, request):
+        except ValueError as e:
+            raise ValidationError(unicode(e))
+
+        return self._get_representation(course, course_key, discussion_settings)
```
The [edX REST doc](https://openedx.atlassian.net/wiki/spaces/OpenDev/pages/18350757/edX+REST+API+Conventions#edXRESTAPIConventions-3.HTTPVerbs) mentions a PATCH request must return 204 No content. Do you think it should be better changing this into a PUT or just return 204?

lgp171188 replied:
@viadanna, this endpoint mimics the behaviour of the existing endpoint, which does the wrong thing. If we just return a 204, an additional GET request is needed to retrieve the updated settings. Is that okay?

lgp171188 replied:
@viadanna, in terms of semantics, a `PUT` would replace the whole object and not do a partial update. And the code for this method also uses the [JSON Merge Patch](https://tools.ietf.org/html/rfc7396) method and its semantics to do a partial update. [Here](https://openedx.atlassian.net/wiki/spaces/OpenDev/pages/18350757/edX%2BREST%2BAPI%2BConventions#edXRESTAPIConventions-AnoteonHTTPPATCH) is what the new API guidelines state about this. So I think it okay to return a `HTTP 204 No Content` response status code to denote success. If the client needs the updated settings (which can be calculated by the client as well), an additional GET request has to be made. Since the data returned for that request will be small enough (it is just a few fields), I think we can ignore the cost of making an additional request.

lgp171188 replied:
Is it okay to return 204 and call this done?

natabene commented:
@lgp171188  Thank you. @marcotuts  Can you have a look at this?

marcotuts commented:
Thanks for this! I will need to sync with @nasthagiri  or another architect to take a look at this and the related WIP API. Thanks for the contribution / proposal here!

lgp171188 commented:
@marcotuts, thanks for the update. The other related PR is complete and it looks like the author (I collaborated with him) forgot to change the title and description to reflect that. Will get it updated.

lgp171188 commented:
@marcotuts, did you get a chance to review this PR?

natabene commented:
@lgp171188 Sorry for delay, we will get to this as soon as we can.

marcotuts commented:
Moving on to arch engineering review next. Thanks!

lgp171188 commented:
@marcotuts, thanks for the update! It looks like the PR has some conflicts due to some recent changes on the master branch. I will resolve those asap and have the PR ready.

natabene commented:
@lgp171188  Please let us know once this is ready for review again.

lgp171188 commented:
@natabene, I have fixed the merge conflicts. This is now ready for review

natabene commented:
@lgp171188  Thank you, i have queued it for review.

natabene commented:
@ormsbee  Any chance you give this a look?

ormsbee commented:
@davestgermain: This is a follow-on to #18975. Do you think you'll have time to review this? I can take a crack at it, but I suspect you have a lot more context.

lgp171188 commented:
jenkins run js

lgp171188 commented:
> This looks fine to me, considering it doesn't change the way discussions work. It just adds an API to manage the existing settings.

> My only minor request is to switch from using `unicode` to `six.text_type`.

@davestgermain, I have updated the PR to address this and squashed, rebased the commits.

edx-webhook commented:
@lgp171188 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/19052&entry.1484655318=https://github.com/lgp171188&entry.752974735=2018-10-03+16:34&entry.1917517419=2019-02-20+15:28&entry.2133058324=Yes) so we can improve your experience in the future.
